src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.DSA.cs

```diff
@@ -115,71 +118,75 @@ protected override bool Sign(
                 Debug.Assert(Helpers.IsDSASupported);
                 signatureParameters = null;
 
-                // If there's no private key, fall back to the public key for a "no private key" exception.
-                DSA? dsa = key as DSA ??
-                    PkcsPal.Instance.GetPrivateKeyForSigning<DSA>(certificate, silent) ??
-                    certificate.GetDSAPublicKey();
-
-                if (dsa == null)
+                using (GetSigningKey(key, certificate, silent, DSACertificateExtensions.GetDSAPublicKey, out DSA? dsa))
```

Comment:
Note: Using `DSACertificateExtensions.GetDSAPublicKey` is more succinct, but the main reason for changing to this is because I was getting errors due to https://github.com/dotnet/roslyn-analyzers/issues/4282.